test/TEST-99-UDEV-STORAGE/test.sh

```diff
@@ -0,0 +1,188 @@
+    printf "    * %s\n" "${passed[@]}"
+    echo "Skipped tests: ${#skipped[@]}"
+    printf "    * %s\n" "${skipped[@]}"
+    echo "Failed tests: ${#failed[@]}"
+    printf "    * %s\n" "${failed[@]}"
+
+    [[ ${#failed[@]} -eq 0 ]] || return 1
+
+    return 0
+}
+
+testcase_megasas2_basic() {
+    if ! "${QEMU_BIN:?}" -device help | grep 'name "megasas-gen2"'; then
+        echo "megasas-gen2 device driver is not available, skipping test..."
+        return 77
+    fi
+
+    local qemu_opts=(
+        "-device megasas-gen2,id=scsi0"
+        "-device megasas-gen2,id=scsi1"
+        "-device megasas-gen2,id=scsi2"
+        "-device megasas-gen2,id=scsi3"
+    )
+
+    for i in {0..127}; do
+        # Add 128 drives, 32 per bus
+        qemu_opts+=(
+            "-device scsi-hd,drive=drive$i,bus=scsi$((i / 32)).0,channel=0,scsi-id=$((i % 32)),lun=0"
+            "-drive format=raw,cache=unsafe,file=${TESTDIR:?}/disk$i.img,if=none,id=drive$i"
+        )
+    done
+
+    KERNEL_APPEND="systemd.setenv=TEST_FUNCTION_NAME=${FUNCNAME[0]} ${USER_KERNEL_APPEND:-}"
+    QEMU_OPTIONS="${qemu_opts[*]} ${USER_QEMU_OPTIONS:-}"
+    test_run_one "${1:?}"
+}
+
+testcase_nvme_basic() {
+    if ! "${QEMU_BIN:?}" -device help | grep 'name "nvme"'; then
+        echo "nvme device driver is not available, skipping test..."
+        return 77
+    fi
+
+    for i in {0..27}; do
+        qemu_opts+=(
+            "-device nvme,drive=nvme$i,serial=deadbeef$i,num_queues=8"
+            "-drive format=raw,cache=unsafe,file=${TESTDIR:?}/disk$i.img,if=none,id=nvme$i"
+        )
+    done
+
+    KERNEL_APPEND="systemd.setenv=TEST_FUNCTION_NAME=${FUNCNAME[0]} ${USER_KERNEL_APPEND:-}"
+    QEMU_OPTIONS="${qemu_opts[*]} ${USER_QEMU_OPTIONS:-}"
+    test_run_one "${1:?}"
+}
+
+# Test for issue https://github.com/systemd/systemd/issues/20212
+testcase_virtio_scsi_identically_named_partitions() {
+    if ! "${QEMU_BIN:?}" -device help | grep 'name "virtio-scsi-pci"'; then
+        echo "virtio-scsi-pci device driver is not available, skipping test..."
+        return 77
+    fi
+
+    local lodev diskpath
+    local qemu_opts=("-device virtio-scsi-pci,id=scsi0,num_queues=4")
+
+    for i in {0..15}; do
+        # Create 16 disks, with 8 partitions per disk (all identically named)
+        # and attach them to a virtio-scsi controller
+        diskpath="${TESTDIR:?}/namedpart$i.img"
+
+        # Save some time (and storage life) during local testing
+        if [[ ! -e "$diskpath" ]]; then
+            dd if=/dev/zero of="$diskpath" bs=1M count=18
+            lodev="$(losetup --show -f -P "$diskpath")"
+            sfdisk "${lodev:?}" <<EOF
+    label: gpt
+
+    name="Hello world", size=2M
+    name="Hello world", size=2M
+    name="Hello world", size=2M
+    name="Hello world", size=2M
+    name="Hello world", size=2M
+    name="Hello world", size=2M
+    name="Hello world", size=2M
+    name="Hello world", size=2M
+EOF
+            losetup -d "$lodev"
+        fi
```

Comment:
Is it necessary to call `sfdisk` for each image? How about to create one partitioned image, and copy it?

Reply:
Why I didn't think of this :facepalm:, thanks!